pkg/resource/reference_resolver.go

```diff
@@ -0,0 +1,222 @@
+*/
+
+package resource
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+)
+
+const (
+	attributeReferencerTagName = "attribute_referencer"
+)
+
+// resourceNotReadyErr is a custom error used for references validation
+type resourceNotReadyErr struct{ msg string }
+
+func newResourceNotReadyErr(format string, args ...interface{}) *resourceNotReadyErr {
+	return &resourceNotReadyErr{fmt.Sprintf(format, args...)}
+}
+
+func (r *resourceNotReadyErr) Error() string {
+	return r.msg
+}
+
+// LocalObjectGetter is a function that is used to retrieve local object references
+type LocalObjectGetter func(ref corev1.LocalObjectReference, obj runtime.Object) error
+
+// AttributeReferencer is an interface for referencing and resolving
+// cross-resource attribute references. See
+// https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-cross-resource-referencing.md
+// for more information
+type AttributeReferencer interface {
+	// Build retrieves and builds the referenced attribute
+	Build(getter LocalObjectGetter) (value string, err error)
+
+	// Assign accepts a resource object, and assigns the given value to the corresponding property
+	Assign(mgd Managed, value string) error
+}
+
+// A ManagedReferenceResolver resolves the references to other managed resources
+type ManagedReferenceResolver interface {
+	ResolveReferences(context.Context, Managed) error
+}
+
+// ReferenceResolver resolves implements ManagedReferenceResolver interface
+type ReferenceResolver struct {
+	reader client.Reader
+}
+
+// NewReferenceResolver returns a new ReferenceResolver
+func NewReferenceResolver(c client.Reader) *ReferenceResolver {
+	return &ReferenceResolver{c}
+}
+
+// ResolveReferences resolves references made to other managed resources
+func (r *ReferenceResolver) ResolveReferences(ctx context.Context, mgd Managed) (err error) {
+	// this recovers from potential panics during execution of
+	// AttributeReferencer methods
+	defer func() {
+		if r := recover(); r != nil {
+			err = errors.Errorf("paniced while resolving references: %v", r)
+		}
+	}()
+
+	referencers, err := findRefResolverFields(mgd, false)
+	if err != nil {
+		return err
+	}
+
+	getter := r.getterFunc(ctx, mgd)
+	err = validateReferencesAreReady(getter, referencers)
+	if err != nil {
+		return err
+	}
+
+	for _, referencer := range referencers {
+		val, err := referencer.Build(getter)
+		if err != nil {
+			return err
+		}
+
+		err = referencer.Assign(mgd, val)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getterFunc returns a function to retrieve local references
+func (r *ReferenceResolver) getterFunc(ctx context.Context, mgd Managed) LocalObjectGetter {
```

Comment:
Nitpick: I see function shortened to `Fn` more commonly than `Func` in Go. The only citation I can provide for this off the top of my head is  https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate